unity-client/Assets/Rendering/Culling/CullingController.cs

```diff
@@ -0,0 +1,505 @@
+        {
+            Material firstMat = renderer.sharedMaterials[0];
+
+            if (firstMat == null)
+                return true;
+
+            if (firstMat.HasProperty(ShaderUtils.ZWrite) &&
+                (int) firstMat.GetFloat(ShaderUtils.ZWrite) == 0)
+            {
+                return false;
+            }
+
+            return true;
+        }
+
+        /// <summary>
+        /// Determines if the given renderer has emissive material traits.
+        /// </summary>
+        /// <param name="renderer">Renderer to be checked.</param>
+        /// <returns>True if the renderer is emissive.</returns>
+        private bool IsEmissive(Renderer renderer)
+        {
+            Material firstMat = renderer.sharedMaterials[0];
+
+            if (firstMat == null)
+                return false;
+
+            if (firstMat.HasProperty(ShaderUtils.EmissionMap) && firstMat.GetTexture(ShaderUtils.EmissionMap) != null)
+                return true;
+
+            if (firstMat.HasProperty(ShaderUtils.EmissionColor) && firstMat.GetColor(ShaderUtils.EmissionColor) != Color.clear)
+                return true;
+
+            return false;
+        }
+
+
+        /// <summary>
+        /// Reset all tracked renderers properties. Needed when toggling or changing settings.
+        /// </summary>
+        void ResetObjects()
+        {
+            foreach (var r in sceneObjects.skinnedRenderers)
+            {
+                r.updateWhenOffscreen = true;
+            }
+
+            foreach (var anim in sceneObjects.animations)
+            {
+                anim.cullingType = AnimationCullingType.AlwaysAnimate;
+            }
+
+            foreach (var r in sceneObjects.renderers)
+            {
+                r.forceRenderingOff = false;
+            }
+        }
+
+        public void Dispose()
+        {
+            Stop();
+        }
+
+        /// <summary>
+        /// Sets the scene objects dirtiness.
+        /// In the next update iteration, all the scene objects are going to be gathered.
+        /// This method has performance impact. 
+        /// </summary>
+        public void SetDirty()
+        {
+            sceneObjects.dirty = true;
+        }
+
+        /// <summary>
+        /// Set settings. This will dirty the scene objects and has performance impact.
+        /// </summary>
+        /// <param name="settings">Settings to be set</param>
+        public void SetSettings(CullingControllerSettings settings)
+        {
+            this.settings = settings;
+            profiles = new List<CullingControllerProfile> {settings.rendererProfile, settings.skinnedRendererProfile};
+            SetDirty();
+        }
+
+        /// <summary>
+        /// Get current settings copy. If you need to modify it, you must set them via SetSettings afterwards.
+        /// </summary>
+        /// <returns>Current settings object copy.</returns>
+        public CullingControllerSettings GetSettings()
```

Comment:
Changed it with a bit of reluctance because it breaks the symmetry of the `Get`/`Set` pairs. Maybe its just better to make the settings object a struct and be done with it. Going to think about this improvement for a later time though.